libs/server/Storage/Session/MainStore/MainStoreOps.cs

```diff
@@ -582,14 +614,35 @@ private unsafe GarnetStatus RENAME(ArgSlice oldKeySlice, ArgSlice newKeySlice, S
                         var expireSpan = new SpanByteAndMemory();
                         var ttlStatus = TTL(ref oldKey, storeType, ref expireSpan, ref context, ref objectContext, true);
 
+                        // Find if this is ETag based key
+                        SpanByte getWithEtagInput = SpanByte.Reinterpret(stackalloc byte[NumUtils.MaximumFormatInt64Length]);
+                        byte* inputPtr = getWithEtagInput.ToPointer();
+                        ((RespInputHeader*)inputPtr)->cmd = RespCommand.GETWITHETAG;
+                        ((RespInputHeader*)inputPtr)->flags = 0;
+
+                        // Check if etag is nil for getWithEtagStatus, this tells us if the key already had an etag associated with it
+                        bool hasEtag = false;
+                        SpanByteAndMemory etagAndDataOutput = new SpanByteAndMemory();
+                        GarnetStatus getWithEtagStatus = GET(ref oldKey, ref getWithEtagInput, ref etagAndDataOutput, ref context);
+                        // we know this key exists, so there is no reason for the status to not be OK
+                        Debug.Assert(getWithEtagStatus == GarnetStatus.OK);
+                        // since we didn't give the etagAndDataOutput span any memory when creating it, the backend would necessarily have had to allocate heap memory if item is not NOTFOUND
+                        Debug.Assert(!etagAndDataOutput.IsSpanByte);
+                        MemoryHandle outputMemHandle = etagAndDataOutput.Memory.Memory.Pin();
+                        byte* outputBufCurr = (byte*)outputMemHandle.Pointer;
+                        byte* end = outputBufCurr + etagAndDataOutput.Length;
+                        RespReadUtils.ReadUnsignedArrayLength(out int numItemInArr, ref outputBufCurr, end);
+                        Debug.Assert(numItemInArr == 2);
+                        hasEtag = !RespReadUtils.ReadNil(ref outputBufCurr, end, out byte? _);
+
                         if (ttlStatus == GarnetStatus.OK && !expireSpan.IsSpanByte)
                         {
                             using var expireMemoryHandle = expireSpan.Memory.Memory.Pin();
                             var expirePtrVal = (byte*)expireMemoryHandle.Pointer;
                             RespReadUtils.TryRead64Int(out var expireTimeMs, ref expirePtrVal, expirePtrVal + expireSpan.Length, out var _);
 
                             // If the key has an expiration, set the new key with the expiration
-                            if (expireTimeMs > 0)
+                            if (expireTimeMs > 0 && !hasEtag)
```

Comment:
put the two existing "if expireTimeMs" blocks inside an if (!hasEtag) then add the etag processing in an else

Reply:
done